### PR TITLE
(Smaller) Construction PR - (IC Construction)

### PR DESCRIPTION
**Pull request by Git-Nivrak**
(Sorry for this being still a big PR next ones will come in smaller batches)
This PR adds:
### Content (The ability to build) 
- Lattice
- Catwalk
- Plating
- Underplating
- Firelock
- Metal Rods
- ~~Reinforced Glass~~ (TODO: Make it craftable in Autolathe and Prolathe)
- Tables

### Code
- SetStackCount component
- FloorTileItem now allows one item to make multiple floors
- BaseTurf fixes

### CODE REVIEW

**Review by SweptWasTaken — APPROVED** [review]

File: Content.Server/GameObjects/Components/Atmos/FirelockComponent.cs
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Atmos/FirelockComponent.cs:72]:
```suggestion
            if (!await base.InteractUsing(eventArgs))
                return false;
```

File: Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs:26]: Should be called count or something as default is vague.

File: Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs:76]: Should this component delete itself if null?

File: Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs:22]: I know it's not part of the PR but do you mind resolving this when needed rather than as a member variable? Given there's a lot of loose tiles on the station it'd be decent memory savings.

File: Content.Shared/Construction/ConstructionConditions/TileType.cs
Comment (metalgearsloth) [inline Content.Shared/Construction/ConstructionConditions/TileType.cs:19]: property rather than field.

File: Content.Shared/GameObjects/Components/SharedStackComponent.cs
Comment (metalgearsloth) [inline Content.Shared/GameObjects/Components/SharedStackComponent.cs:131]: Someone needs to refactor this so bad. Fine for PR.

File: Content.Shared/Maps/TurfHelpers.cs
Comment (metalgearsloth) [inline Content.Shared/Maps/TurfHelpers.cs:14]:
```suggestion
```

File: Resources/Textures/Objects/Materials/sheets.rsi/meta.json
Comment (metalgearsloth) [inline Resources/Textures/Objects/Materials/sheets.rsi/meta.json:187]: Don't need delays if it's just 1

Also is the licence for these the same as the other states?

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
There's an empty plating file. Get Zum to approve construction.

I also assume it all works in game but I should probably check it.

File: Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs:33]: You should just take the first tile in the list as the default, not do this ugly hack.

File: Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs:79]: Why would you snowflake-check one tile first, then run the for loop to check the rest? Just do the for loop and always prioritize the first element of the list.

File: Content.Shared/Construction/ConstructionConditions/TileType.cs
Comment (Zumorica) [inline Content.Shared/Construction/ConstructionConditions/TileType.cs:23]: Please don't default to an actual tile. Default to an empty string.

File: Resources/Prototypes/Entities/Constructible/Doors/firelock_frame.yml
Comment (Zumorica) [inline Resources/Prototypes/Entities/Constructible/Doors/firelock_frame.yml:3]: Incorrect name?

File: Resources/Prototypes/Entities/Constructible/Items/metal_rod.yml
Comment (Zumorica) [inline Resources/Prototypes/Entities/Constructible/Items/metal_rod.yml:20]: Default to 50.

File: Resources/Prototypes/Entities/Constructible/Items/metal_rod.yml
Comment (Zumorica) [inline Resources/Prototypes/Entities/Constructible/Items/metal_rod.yml:4]: Make this has a "full" suffix after you address my review below.
Also make another prototype that inherits this one that only has 1 metal rod in the stack and give that the "1" suffix or something. Check how other sheets/materials did it for consistency.

File: Resources/Prototypes/Entities/Objects/Misc/firelock_electronics.yml
Comment (Zumorica) [inline Resources/Prototypes/Entities/Objects/Misc/firelock_electronics.yml:5]:
```suggestion
  description: Advanced circuit board used to detect differences in pressure, temperature and gas concentrations between the two sides of the door.
```

File: Resources/Prototypes/Recipes/Construction/Graphs/catwalk.yml
Comment (Zumorica) [inline Resources/Prototypes/Recipes/Construction/Graphs/catwalk.yml:10]:
```suggestion
              amount: 2
```

File: Resources/Prototypes/Recipes/Construction/Graphs/catwalk.yml
Comment (Zumorica) [inline Resources/Prototypes/Recipes/Construction/Graphs/catwalk.yml:19]:
```suggestion
                amount: 2
```

File: Resources/Prototypes/Recipes/Construction/Graphs/catwalk.yml
Comment (Zumorica) [inline Resources/Prototypes/Recipes/Construction/Graphs/catwalk.yml:7]: Give this edge a completion action that snaps the entity to the grid.

File: Resources/Prototypes/Recipes/Construction/Graphs/firelock.yml
Comment (Zumorica) [inline Resources/Prototypes/Recipes/Construction/Graphs/firelock.yml:16]: Give this node an action that changes the sprite state to frame1. For reversing the construction.

File: Resources/Prototypes/Recipes/Construction/Graphs/firelock.yml
Comment (Zumorica) [inline Resources/Prototypes/Recipes/Construction/Graphs/firelock.yml:54]: Prototype steps are discouraged. Please make a "dummy" component called `FirelockElectronicsComponent`, give that to the firelock electronics prototype and use a component step instead.

File: Resources/Prototypes/Recipes/Construction/Graphs/firelock.yml
Comment (Zumorica) [inline Resources/Prototypes/Recipes/Construction/Graphs/firelock.yml:71]: Shouldn't this be frame3?

File: Resources/Prototypes/Recipes/Construction/Graphs/firelock.yml
Comment (Zumorica) [inline Resources/Prototypes/Recipes/Construction/Graphs/firelock.yml:130]: This means you can add glass to ANY regular firelocks to give them windows... I like the idea of constructing window firelocks, but that edge should be in other node, not here. You could add a new node, and an edge to it in frame4.

File: Resources/Prototypes/Recipes/Construction/Graphs/tables.yml
Comment (Zumorica) [inline Resources/Prototypes/Recipes/Construction/Graphs/tables.yml:18]: I don't really like this edge... Why don't you make an edge from the wood table node to the poker table node?

File: Resources/Prototypes/Recipes/Construction/structures.yml
Comment (Zumorica) [inline Resources/Prototypes/Recipes/Construction/structures.yml:135]: Given this is a table *frame* perhaps it should have a different description.

File: Resources/Prototypes/Recipes/Construction/structures.yml
Comment (Zumorica) [inline Resources/Prototypes/Recipes/Construction/structures.yml:166]: You should be able to build this on top of tiles other than lattice. (See our maint: We use these and they look cool. Wires go under it.)

File: Resources/Prototypes/Recipes/Construction/structures.yml
Comment (Zumorica) [inline Resources/Prototypes/Recipes/Construction/structures.yml:163]: Doesn't really apply to our codebase right now...

File: Resources/Textures/Objects/Materials/sheets.rsi/meta.json
Comment (Git-Nivrak) [reply Resources/Textures/Objects/Materials/sheets.rsi/meta.json:187]: No idea ask swept

File: Resources/Prototypes/Tiles/floors.yml
Comment (Zumorica) [inline Resources/Prototypes/Tiles/floors.yml:148]: All these tiles (in the file) should have plating, not underplating.

**Review by Zumorica — CHANGES_REQUESTED** [review]

File: Resources/Prototypes/Recipes/Construction/structures.yml
Comment (Git-Nivrak) [reply Resources/Prototypes/Recipes/Construction/structures.yml:135]: ~~Technically it's still a square piece of metal standing on four lags~~ Sure how about "A sturdy frame used in the construction of tables"

**Comment by Git-Nivrak** [comment]
I think that should be all unless I forgot something

File: Content.Server/Construction/Conditions/DoorWelded.cs
Comment (Zumorica) [inline Content.Server/Construction/Conditions/DoorWelded.cs:26]: This should check that `IsWeldedShut` is the same as `Welded`, right?

File: Content.Server/Construction/Conditions/DoorWelded.cs
Comment (Zumorica) [inline Content.Server/Construction/Conditions/DoorWelded.cs]: You should have a message for when `Welded` is false.

File: Content.Server/GameObjects/Components/Atmos/FirelockComponent.cs
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Atmos/FirelockComponent.cs]: Instead of checking this twice... Encapsulare this in a `if (!IsWeldedShut)` check.

File: Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs]: You're not doing this correctly... Both this else if and the if above should be part of the for loop for all output tiles, and break when you correctly place a tile.

File: Resources/Prototypes/Recipes/Construction/Graphs/APC.yml
Comment (Zumorica) [inline Resources/Prototypes/Recipes/Construction/Graphs/APC.yml:13]: I assume this is a wip, yes?

File: Resources/Prototypes/Recipes/Construction/structures.yml
Comment (Zumorica) [inline Resources/Prototypes/Recipes/Construction/structures.yml:135]:
```suggestion
  description: A sturdy frame used in the construction of tables
```

**Comment by Zumorica** [comment]
Reviewed again. Just a few more fixes and it should be good to go. :+1:

File: Resources/Prototypes/Recipes/Construction/Graphs/APC.yml
Comment (Git-Nivrak) [reply Resources/Prototypes/Recipes/Construction/Graphs/APC.yml:13]: Yes

File: Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs:70]: Maybe move getting the grid outside the foreach loop

File: Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs:78]: This can never be null here

File: Content.Server/GameObjects/Components/Atmos/FirelockComponent.cs
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Atmos/FirelockComponent.cs:90]: mmmm, yeah. Maybe add this condition to the if check above instead of here. That wait you won't get the do_After when the firelock is obviously welded.

File: Content.Server/Construction/Conditions/DoorWelded.cs
Comment (Zumorica) [inline Content.Server/Construction/Conditions/DoorWelded.cs]: Mmm, this is wrong.
If the door is welded, and we need the door to be unwelded it should say "First, unweld the door."
If the door is unwelded, and we need the door to be welded it should say "First, weld the door."

**Review by Zumorica — CHANGES_REQUESTED** [review]

File: Content.Server/GameObjects/Components/Atmos/FirelockComponent.cs
Comment (Git-Nivrak) [reply Content.Server/GameObjects/Components/Atmos/FirelockComponent.cs:90]: Not sure exactly what you mean here

Comment (Zumorica) [reply Content.Server/GameObjects/Components/Atmos/FirelockComponent.cs:90]: Instead of doing this if check just after the one before, you can merge both. So it'd check if it's welded shut && then the do after

File: Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs
Comment (Git-Nivrak) [reply Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs:22]: No idea how to do that

**Review by Clyybber — APPROVED** [review]

File: Content.Server/GameObjects/Components/Atmos/FirelockComponent.cs
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Atmos/FirelockComponent.cs]: this is still incorrect! This is what I meant...
```
                if (IsWeldedShut || !await tool.UseTool(eventArgs.User, Owner, holdingPressure || holdingFire ? 1.5f : 0.25f, ToolQuality.Prying)) return false;
                if (State == DoorState.Closed)
                {
                    Open();
                }else if (State == DoorState.Open)
                {
                    Close();
```

**Comment by Git-Nivrak** [comment]
Well
This is not for no reason, For some reason (I honestly don't know why) it doesn't work without it

**Comment by Zumorica** [comment]
Let me check out your branch and test

**Comment by Zumorica** [comment]
I checked, and you made a mistake. The base InteractUsing method call is negated, when it shouldn't be.
I was going to push a fix to your branch, but seems you've disabled edits from maintainers. Here's how the entire method should look:

```csharp
public override async Task<bool> InteractUsing(InteractUsingEventArgs eventArgs)
        {
            if (await base.InteractUsing(eventArgs))
                return false;

            if (!eventArgs.Using.TryGetComponent<ToolComponent>(out var tool))
                return false;

            if (tool.HasQuality(ToolQuality.Prying) && !IsWeldedShut)
            {
                var holdingPressure = IsHoldingPressure();
                var holdingFire = IsHoldingFire();

                if (State == DoorState.Closed)
                {
                    if(holdingPressure)
                        Owner.PopupMessage(eventArgs.User, "A gush of air blows in your face... Maybe you should reconsider.");
                }

                if (IsWeldedShut || !await tool.UseTool(eventArgs.User, Owner, holdingPressure || holdingFire ? 1.5f : 0.25f, ToolQuality.Prying)) return false;
                if (State == DoorState.Closed)
                {
                    Open();
                }else if (State == DoorState.Open)
                {
                    Close();
                }


                return true;
            }


            return false;
        }
```

**Comment by Git-Nivrak** [comment]
Done

**Review by Zumorica — APPROVED** [review]

**Comment by Git-Nivrak** [comment]
Pog

**Pull request merged**
